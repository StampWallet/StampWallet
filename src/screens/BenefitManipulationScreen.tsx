import React, { useEffect } from 'react';
import { StyleSheet, StatusBar, View, Platform, ScrollView, SafeAreaView } from 'react-native';
import { useForm, FormProvider } from 'react-hook-form';

import StyleBase from '../styles/StyleBase';

import TopBar from '../components/Bars/TopBar';
import HookFormImagePicker from '../components/HookFormComponents/HookFormImagePicker';
import HookFormDatePicker from '../components/HookFormComponents/HookFormDatePicker';
import CustomButton from '../components/Miscellaneous/CustomButton';
import BoxContainer from '../components/Miscellaneous/BoxContainer';
import HookFormInput from '../components/HookFormComponents/HookFormInput';

import { required, validateNumber } from '../utils/validators';
import { MAX_DATE, MIN_DATE } from '../constants/numericAndStringConstants';

/*
todo: 
      integrate/implement datepicker
      check if existing or new benefit
      button onPress sends request through api
      fix scroll view
      check if noBenefits integer
      check maxAmount
      provide initial value to forminput
*/

export default function BenefitManipulationScreen({ navigation, route }) {
  const name = /*Benefits.name*/ '';
  const description = /*Benefits.description*/ '';
  const maxAmount = /*Benefits.maxAmount*/ '';

  const data = route?.params?.params;

  console.log(data);

  const { ...methods } = useForm();
  const {
    handleSubmit,
    control,
    formState: { errors },
    watch,
    setValue,
  } = methods;

  const pointsWatch = watch('noPoints');
  const benefitsWatch = watch('noBenefits');
  const dateFromWatch = watch('dateFrom');
  const dateToWatch = watch('dateToWatch');

  const title = data /*.publicId*/ ? 'Edit benefit' : 'Create benefit';
  const onPress = data /*.publicId*/
    ? handleSubmit(() => alert('Work on edition in progress'))
    : handleSubmit(() => alert('Work on creation in progress'));

  useEffect(() => {
    if (!data) {
      return;
    }

    setValue('name', data.name);
    setValue('price', data.price);
    setValue('desc', data.description);
    setValue('dateFrom', data.startDate);
    // setValue('benefits', data.maxAmount);
  }, [data, setValue]);

  return (
    <SafeAreaView style={StyleBase.container}>
      <StatusBar barStyle='default' />
      <TopBar iconLeft='arrow-left' onPressLeft={() => navigation.pop()} />
      <ScrollView
        style={styles.scrollView}
        contentContainerStyle={{ marginBottom: 10, justifyContent: 'center', alignItems: 'center' }}
      >
        <FormProvider {...methods}>
          <BoxContainer style={styles.boxContainer}>
            {/* show current value? */}
            <HookFormInput
              control={control}
              rules={{ required }}
              placeholder='name'
              name='name'
              isInvalid={Boolean(errors.name)}
            />
            <HookFormInput
              control={control}
              rules={{ required, validate: () => validateNumber(pointsWatch) }}
              placeholder='number of points required'
              name='points'
              isInvalid={Boolean(errors.points)}
            />
            <HookFormInput
              control={control}
              rules={{ required }}
              placeholder='short description'
              name='desc'
              isInvalid={Boolean(errors.desc)}
            />
            <HookFormImagePicker
              control={control}
              rules={{ required }}
              name='benefitIcon'
              isInvalid={Boolean(errors.benefitIcon)}
            />
            <View style={styles.containerDatepicker}>
              <View style={styles.datePicker}>
                <HookFormDatePicker
                  control={control}
                  name='dateFrom'
                  rules={{ required }}
                  minDate={MIN_DATE}
                  maxDate={dateToWatch || MAX_DATE}
                  isInvalid={Boolean(errors.dateFrom)}
                />
              </View>
              <View style={styles.datePicker}>
                <HookFormDatePicker
                  control={control}
                  name='dateTo'
                  rules={{ required }}
                  minDate={dateFromWatch || MIN_DATE}
                  maxDate={MAX_DATE}
                  isInvalid={Boolean(errors.dateTo)}
                />
              </View>
            </View>
            <HookFormInput
              control={control}
              rules={{ required, validate: () => validateNumber(benefitsWatch) }}
              placeholder='maximum number of benefits in inventory'
              name='benefits'
              isInvalid={Boolean(errors.benefits)}
            />
            <CustomButton title={title} onPress={onPress} customButtonStyle={{ width: '100%' }} />
          </BoxContainer>
        </FormProvider>
      </ScrollView>
    </SafeAreaView>
  );
}

const styles = StyleSheet.create({
  scrollView: {
    //backgroundColor: '#000',
    marginTop: '15%',
    height: '80%',
    width: 350,
  },
  containerDatepicker: {
    height: 100,
    width: '90%',
    flexDirection: 'row',
    alignItems: 'center',
    justifyContent: 'flex-start',
    marginBottom: 10,
    //backgroundColor: '#000',
  },
  boxContainer: {
    height: '80%',
    width: 350,
    alignItems: 'center',
    justifyContent: 'center',
  },
  text: {
    fontFamily: Platform.select({ ios: 'Arial', android: 'Roboto' }),
    fontSize: 24,
    fontWeight: '200',
    textDecorationLine: 'underline',
  },
  datePicker: {
    width: '50%',
    justifyContent: 'center',
    alignItems: 'center',
  },
});
