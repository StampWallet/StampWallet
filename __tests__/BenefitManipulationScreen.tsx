import React from 'react';
import { describe, expect } from '@jest/globals';
import { render, fireEvent } from 'react-native-testing-library';
import '@testing-library/jest-native/extend-expect';

import BenefitManipulationScreen from '../screens/BenefitManipulationScreen';

describe('Benefit manipulation tests', () => { 
    it('checks if rendered correctly', () => {
        const tree = render(<BenefitManipulationScreen />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it('checks content of the screen', () => {
        const tree = render(<BenefitManipulationScreen />);
        expect(tree.getAllByTestId('Text input')).toHaveLength(6);
        expect(tree.getAllByTestId('Text')).toHaveLength(1);
        expect(tree.getByTestId('dropFile')).toBeVisible;
        expect(tree.getByTestId('Button')).toBeVisible;

    })

    it('checks if error is shown', () => {
        let tree = render(<BenefitManipulationScreen />);
        fireEvent.changeText(tree.getByTestId('Text input'), "");
        fireEvent.press(tree.getByTestId('Button'));
        expect(tree.getByText('This field is required')).toBeVisible;
    })

    it('checks if Benefit is going to be saved', () => {
        const addBenefit = jest.fn();
        const date = new Date()
        const args = ['Name', 100, 'Description', date, (new Date()).setDate(date.getDate() + 1), 2];
        let tree = render(<BenefitManipulationScreen />);
        let textInputs = tree.getAllByTestId('Text input');
        for(let i = 0; i < textInputs.length; i++) { fireEvent.changeText(textInputs[i]), args[i] };
        fireEvent.press(tree.getByTestId('Button'));
        expect(addBenefit).toBeCalled;
    })
 })