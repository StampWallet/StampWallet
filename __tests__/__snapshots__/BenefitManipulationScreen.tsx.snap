// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Benefit manipulation tests checks if rendered correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#fff",
      "flex": 1,
      "justifyContent": "center",
    }
  }
>
  <Text>
    This is benefit addition and edition screen
  </Text>
</View>
`;
